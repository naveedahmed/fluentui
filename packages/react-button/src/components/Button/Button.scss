@import '~@fluentui/common-styles/dist/sass/focusBorder';

/**
  * Slot definitions
  *
  * Slot definitions must align to the names of the slots themselves. They should:
  *
  * 1. Contain the actual css property definitions (E.g. background-color: {value})
  * 2. Property definitions should map to tokens using the format: {component}-{property}
  * 3. Avoid hard coding values unless there are absolutely no use cases to change
  *    the property value.
  * 4. Order hardcoded css values first, tokens based second. This way we can identify
  *    what is hardcoded easily.
 **/

.root {
  cursor: pointer;
  align-items: center;
  border-style: solid;
  display: inline-flex;
  justify-content: center;
  outline: none;
  position: relative;
  user-select: none;
  box-sizing: border-box;
  vertical-align: middle;

  background: var(--button-background, white);
  color: var(--button-contentColor, #2c2621);

  border-color: var(--button-borderColor);
  border-radius: var(--button-borderRadius);
  border-width: var(--button-borderWidth);
  box-shadow: var(--button-boxShadow);

  width: var(--button-width);
  height: var(--button-height);
  min-height: var(--button-minHeight);

  cursor: pointer;
  padding: var(--button-padding);
  transition: var(--button-transition);
  white-space: var(--button-whiteSpace, nowrap);

  font-family: var(--button-fontFamily);
  font-size: var(--button-fontSize);
  font-weight: var(--button-fontWeight);

  @include focus-border(
    $padding: -1px,
    $color: var(--button-focusColor),
    $thickness: var(--button-focusWidth, 2px),
    $radius: var(--button-borderRadius)
  );
  @include focus-box-shadow($hOffset: 0, $vOffset: 0, $blur: 0, $thickness: 1px, $color: var(--button-focusInnerColor));

  /**
  * Modifiers
  *
  * Modifiers must align to the props on the component. They should:
  *
  * 1. Redefine the values for {component}-{property} tokens.
  * 2. Map property definitions to variables using the format: {component}-{property}
  * 3. Avoid defining styles directly.
 **/
  &._primary {
    --button-background: var(--accent-background);
    --button-borderColor: var(--accent-borderColor);
    --button-contentColor: var(--accent-contentColor);
    --button-iconColor: var(--accent-iconColor);

    --button-hovered-background: var(--accent-hovered-background);
    --button-hovered-borderColor: var(--accent-hovered-borderColor);
    --button-hovered-contentColor: var(--accent-hovered-contentColor);
    --button-hovered-iconColor: var(--accent-hovered-iconColor);

    --button-focused-background: var(--accent-focused-background);
    --button-focused-borderColor: var(--accent-focused-borderColor);
    --button-focused-contentColor: var(--accent-focused-contentColor);
    --button-focused-iconColor: var(--accent-focused-iconColor);

    --button-pressed-background: var(--accent-pressed-background);
    --button-pressed-borderColor: var(--accent-pressed-borderColor);
    --button-pressed-contentColor: var(--accent-pressed-contentColor);
    --button-pressed-iconColor: var(--accent-pressed-iconColor);

    --button-disabled-background: var(--accent-disabled-background);
    --button-disabled-borderColor: var(--accent-disabled-borderColor);
    --button-disabled-contentColor: var(--accent-disabled-contentColor);
    --button-disabled-iconColor: var(--accent-disabled-iconColor);
  }

  &._ghost {
    --button-background: var(--ghost-background);
    --button-borderColor: var(--ghost-borderColor);
    --button-contentColor: var(--ghost-contentColor);
    --button-iconColor: var(--ghost-iconColor);

    --button-disabled-background: var(--ghost-disabled-background);
    --button-disabled-borderColor: var(--ghost-disabled-borderColor);
    --button-disabled-contentColor: var(--ghost-disabled-contentColor);
    --button-disabled-iconColor: var(--ghost-disabled-iconColor);

    --button-focused-background: var(--ghost-focused-background);
    --button-focused-borderColor: var(--ghost-focused-borderColor);
    --button-focused-contentColor: var(--ghost-focused-contentColor);
    --button-focused-iconColor: var(--ghost-focused-iconColor);

    --button-hovered-background: var(--ghost-hovered-background);
    --button-hovered-borderColor: var(--ghost-hovered-borderColor);
    --button-hovered-contentColor: var(--ghost-hovered-contentColor);
    --button-hovered-iconColor: var(--ghost-hovered-iconColor);

    --button-pressed-background: var(--ghost-pressed-background);
    --button-pressed-borderColor: var(--ghost-pressed-borderColor);
    --button-pressed-contentColor: var(--ghost-pressed-contentColor);
    --button-pressed-iconColor: var(--ghost-pressed-iconColor);
  }

  &._iconOnly {
    --button-width: var(--button-height, var(--button-minHeight));
    --button-padding: 0;
  }

  &._circular {
    --button-borderRadius: 50000px;
  }

  &._fluid {
    width: 100%;
    max-width: 100%;
  }

  // TODO: confirm if we should support size prop.
  &._size_smallest {
    --button-minHeight: var(--button-size-smallest);
  }

  &._size_smaller {
    --button-minHeight: var(--button-size-smaller);
  }

  &._size_small {
    --button-minHeight: var(--button-size-small);
  }

  &._size_large {
    --button-minHeight: var(--button-size-large);
  }

  &._size_larger {
    --button-minHeight: var(--button-size-larger);
  }

  &._size_largest {
    --button-minHeight: var(--button-size-largest);
  }

  &._disabled {
    pointer-events: none;
    opacity: var(--button-disabled-opacity);
    background-color: var(--button-disabled-background);
    color: var(--button-disabled-contentColor);
    border-color: var(--button-disabled-borderColor);
    box-shadow: var(--button-disabled-boxShadow);
  }
}

.icon {
  color: var(--button-iconColor);
  display: flex;
  align-items: center;
  justify-content: center;
  width: var(--button-iconSize);
  height: var(--button-iconSize);
  font-size: var(--button-iconSize, inherit);
  font-weight: normal;

  ._disabled & {
    color: var(--button-disabled-iconColor);
  }
}

/* selectors */

.root > *:not(:first-child) {
  margin-left: var(--button-contentGap);
}

.root:hover {
  background: var(--button-hovered-background, var(--button-background));
  color: var(--button-hovered-contentColor, var(--button-contentColor));
  border-color: var(--button-hovered-borderColor, var(--button-borderColor));
  box-shadow: var(--button-hovered-boxShadow, var(--button-boxShadow));

  .icon {
    color: var(--button-hovered-iconColor, var(--button-iconColor));
  }
}

.root:active {
  background: var(--button-pressed-background, var(--button-hovered-background));
  color: var(--button-pressed-contentColor, var(--button-hovered-contentColor, var(--button-contentColor)));
  border-color: var(--button-pressed-borderColor, var(--button-hovered-borderColor, var(--button-borderColor)));
  box-shadow: var(--button-pressed-boxShadow, var(--button-hovered-boxShadow, var(--button-boxShadow)));

  transform: var(--button-pressed-transform);
  transition: var(--button-pressed-transition);

  .icon {
    color: var(--button-pressed-iconColor, var(--button-iconColor));
  }
}
