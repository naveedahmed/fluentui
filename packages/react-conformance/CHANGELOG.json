{
  "name": "@fluentui/react-conformance",
  "entries": [
    {
      "date": "Mon, 31 Aug 2020 12:25:46 GMT",
      "tag": "@fluentui/react-conformance_v0.1.1",
      "version": "0.1.1",
      "comments": {
        "patch": [
          {
            "comment": "Updating doctrine and modifying merge function to use lodash.",
            "author": "esteban.230@hotmail.com",
            "commit": "268eff06b3e40b5d965f2187bbb1cd2110c6cb67",
            "package": "@fluentui/react-conformance"
          }
        ]
      }
    }
  ]
}
