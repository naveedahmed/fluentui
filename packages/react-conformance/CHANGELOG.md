# Change Log - @fluentui/react-conformance

This log was last generated on Mon, 31 Aug 2020 12:25:46 GMT and should not be manually modified.

<!-- Start content -->

## [0.1.1](https://github.com/microsoft/fluentui/tree/@fluentui/react-conformance_v0.1.1)

Mon, 31 Aug 2020 12:25:46 GMT

### Patches

- Updating doctrine and modifying merge function to use lodash. ([PR #14344](https://github.com/microsoft/fluentui/pull/14344) by esteban.230@hotmail.com)
