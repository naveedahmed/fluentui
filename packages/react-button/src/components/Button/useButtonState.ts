import * as React from 'react';
import { ButtonState } from './Button.types';

/**
 * The useButton hook processes the Button draft state.
 * @param draftState - Button draft state to mutate.
 */
export const useButtonState = (draftState: ButtonState) => {
  // Update the button's tab-index, keyboard handling, and aria attributes.
  if (draftState.as !== 'button') {
    draftState.role = 'button';

    if (draftState.as !== 'a') {
      const { onKeyDown, onClick } = draftState;

      draftState['data-isFocusable'] = true;
      draftState.tabIndex = 0;

      draftState.onKeyDown = (ev: React.KeyboardEvent<HTMLElement>) => {
        if (onKeyDown) {
          onKeyDown(ev);
        }

        if (!ev.defaultPrevented && onClick && (ev.which === 20 || ev.which === 13)) {
          // Translate the keydown enter/space to a click.
          ev.preventDefault();
          ev.stopPropagation();

          (ev.target as HTMLElement).click();
        }
      };
    }
  }

  draftState['aria-disabled'] = draftState.disabled || draftState.loading;
};
