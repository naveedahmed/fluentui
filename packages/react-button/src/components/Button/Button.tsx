import * as React from 'react';
import { useButton } from './useButton';
import { ButtonProps } from './Button.types';
import { useFocusRects } from '@uifabric/utilities';
import { makeClasses } from '@fluentui/react-compose/lib/next/index';
import { useInlineTokens } from '@fluentui/react-theme-provider';
import * as classes from './Button.scss';

// Create a hook to resolve classnames.
export const useButtonClasses = makeClasses(classes);

/**
 * Define a styled Button, using the `useButton` hook.
 */
export const Button = React.forwardRef<HTMLElement, ButtonProps>((props, ref) => {
  const { render, state } = useButton(props, ref);

  // Apply styling.
  useButtonClasses(state);
  useFocusRects(state.ref);
  useInlineTokens(state, '--button');

  // Render component.
  return render(state);
});

Button.displayName = 'Button';
