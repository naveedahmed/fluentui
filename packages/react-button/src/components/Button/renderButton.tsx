import * as React from 'react';
import { getSlots } from '@fluentui/react-compose/lib/next/index';
import { ButtonState } from './Button.types';
import { buttonShorthandProps } from './useButton';

/**
 * Define the render function. Given the state of a button, renders it.
 */
export const renderButton = (state: ButtonState) => {
  const { slots, slotProps } = getSlots(state, buttonShorthandProps);
  const { loading, iconPosition, iconOnly } = state;

  return (
    <slots.root {...slotProps.root}>
      {loading && <slots.loader {...slotProps.loader} />}
      {iconPosition !== 'after' && <slots.icon {...slotProps.icon} />}
      {!iconOnly && <slots.children {...slotProps.children} />}
      {iconPosition === 'after' && <slots.icon {...slotProps.icon} />}
    </slots.root>
  );
};
