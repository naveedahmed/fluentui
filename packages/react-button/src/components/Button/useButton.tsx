import * as React from 'react';
import { mergeProps, resolveShorthandProps } from '@fluentui/react-compose/lib/next/index';
import { ButtonProps } from './Button.types';
import { useButtonState } from './useButtonState';
import { renderButton } from './renderButton';

/**
 * Consts listing which props are shorthand props.
 */
export const buttonShorthandProps = ['icon', 'loader', 'children'];

/**
 * Given user props, returns state and render function for a Button.
 */
export const useButton = (props: ButtonProps, ref: React.Ref<HTMLElement>, defaultProps?: ButtonProps) => {
  // Ensure that the `ref` prop can be used by other things (like useFocusRects) to refer to the root.
  // NOTE: We are assuming refs should not mutate to undefined. Either they are passed or not.
  // eslint-disable-next-line react-hooks/rules-of-hooks
  const resolvedRef = ref || React.useRef();
  const state = mergeProps(
    {
      ref: resolvedRef,
      as: 'button',
      icon: { as: 'span' },
      children: { as: 'span' },
      loader: { as: 'span' },
    },
    defaultProps,
    resolveShorthandProps(props, buttonShorthandProps),
  );

  useButtonState(state);

  return { state, render: renderButton };
};
