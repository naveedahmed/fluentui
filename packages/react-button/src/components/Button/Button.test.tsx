import * as React from 'react';
import * as path from 'path';
import { isConformant } from '@fluentui/react-conformance';
import { Button } from './Button';
import * as renderer from 'react-test-renderer';
import { mount, ReactWrapper } from 'enzyme';

describe('Button', () => {
  let wrapper: ReactWrapper | undefined;

  afterEach(() => {
    if (wrapper) {
      wrapper.unmount();
      wrapper = undefined;
    }
  });

  isConformant({
    componentPath: path.join(__dirname, 'Button.tsx'),
    Component: Button,
    displayName: 'Button',
    disabledTests: ['has-docblock', 'as-renders-html', 'as-passes-as-value', 'as-renders-react-class', 'as-renders-fc'],
  });

  /**
   * Note: see more visual regression tests for Button in /apps/vr-tests.
   */
  it('renders a default state', () => {
    const component = renderer.create(<Button>Default button</Button>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('can be focused', () => {
    const rootRef = React.createRef<HTMLButtonElement>();

    wrapper = mount(<Button ref={rootRef}>Focus me</Button>);

    expect(typeof rootRef.current).toEqual('object');
    expect(document.activeElement).not.toEqual(rootRef.current);

    rootRef.current?.focus();

    expect(document.activeElement).toEqual(rootRef.current);
  });
});
