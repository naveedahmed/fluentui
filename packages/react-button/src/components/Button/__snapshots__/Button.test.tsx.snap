// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders a default state 1`] = `
<button
  className=""
>
  <span>
    Default button
  </span>
</button>
`;
